File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-disabled-tests.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-disabled-tests', {} as never, {
  valid: [
    { code: 'describe("foo", function () {})' },
    { code: 'it("foo", function () {})' },
    { code: 'describe.only("foo", function () {})' },
    { code: 'it.only("foo", function () {})' },
    { code: 'it.each("foo", () => {})' },
    { code: 'it.concurrent("foo", function () {})' },
    { code: 'test("foo", function () {})' },
    { code: 'test.only("foo", function () {})' },
    { code: 'test.concurrent("foo", function () {})' },
    { code: 'describe[`${"skip"}`]("foo", function () {})' },
    { code: 'it.todo("fill this later")' },
    { code: 'var appliedSkip = describe.skip; appliedSkip.apply(describe)' },
    { code: 'var calledSkip = it.skip; calledSkip.call(it)' },
    { code: '({ f: function () {} }).f()' },
    { code: '(a || b).f()' },
    { code: 'itHappensToStartWithIt()' },
    { code: 'testSomething()' },
    { code: 'xitSomethingElse()' },
    { code: 'xitiViewMap()' },
    {
      code: `
        import { pending } from "actions"

        test("foo", () => {
          expect(pending()).toEqual({})
        })
      `,
    },
    {
      code: `
        const { pending } = require("actions")

        test("foo", () => {
          expect(pending()).toEqual({})
        })
      `,
    },
    {
      code: `
        test("foo", () => {
          const pending = getPending()
          expect(pending()).toEqual({})
        })
      `,
    },
    {
      code: `
        test("foo", () => {
          expect(pending()).toEqual({})
        })

        function pending() {
          return {}
        }
      `,
    },
    {
      code: `
        import { test } from './test-utils';

        test('something');
      `,
    },
  ],
  invalid: [
    {
      code: 'describe.skip("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'describe.skip.each([1, 2, 3])("%s", (a, b) => {});',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xdescribe.each([1, 2, 3])("%s", (a, b) => {});',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'describe[`skip`]("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'describe["skip"]("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'it.skip("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'it["skip"]("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'test.skip("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'it.skip.each``("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'test.skip.each``("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'it.skip.each([])("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'test.skip.each([])("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'test["skip"]("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xdescribe("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xit("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xtest("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xit.each``("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xtest.each``("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xit.each([])("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'xtest.each([])("foo", function () {})',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'it("has title but no callback")',
      errors: [{ message: 'Test is missing function argument' }],
    },
    {
      code: 'test("has title but no callback")',
      errors: [{ message: 'Test is missing function argument' }],
    },
    {
      code: 'it("contains a call to pending", function () { pending() })',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'pending();',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: 'describe("contains a call to pending", function () { pending() })',
      errors: [{ message: 'Tests should not be skipped' }],
    },
    {
      code: `
        import { test } from '@jest/globals';

        test('something');
      `,
      errors: [{ message: 'Test is missing function argument' }],
    },
  ],
});

File: internal/plugins/jest/rules/no_disabled_tests/no_disabled_tests_test.go
package no_disabled_tests_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_disabled_tests"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoDisabledTestsRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_disabled_tests.NoDisabledTestsRule,
		[]rule_tester.ValidTestCase{
			{Code: `describe("foo", function () {})`},
			{Code: `it("foo", function () {})`},
			{Code: `describe.only("foo", function () {})`},
			{Code: `it.only("foo", function () {})`},
			{Code: `it.each("foo", () => {})`},
			{Code: `it.concurrent("foo", function () {})`},
			{Code: `test("foo", function () {})`},
			{Code: `test.only("foo", function () {})`},
			{Code: `test.concurrent("foo", function () {})`},
			{Code: "describe[`${\"skip\"}`](\"foo\", function () {})"},
			{Code: `it.todo("fill this later")`},
			{Code: `var appliedSkip = describe.skip; appliedSkip.apply(describe)`},
			{Code: `var calledSkip = it.skip; calledSkip.call(it)`},
			{Code: `({ f: function () {} }).f()`},
			{Code: `(a || b).f()`},
			{Code: `itHappensToStartWithIt()`},
			{Code: `testSomething()`},
			{Code: `xitSomethingElse()`},
			{Code: `xitiViewMap()`},
			{Code: `
				import { pending } from "actions"

				test("foo", () => {
				  expect(pending()).toEqual({})
				})
			`},
			{Code: `
				const { pending } = require("actions")

				test("foo", () => {
				  expect(pending()).toEqual({})
				})
			`},
			{Code: `
				test("foo", () => {
				  const pending = getPending()
				  expect(pending()).toEqual({})
				})
			`},
			{Code: `
				test("foo", () => {
				  expect(pending()).toEqual({})
				})

				function pending() {
				  return {}
				}
			`},
			{
				Code: `
					import { test } from './test-utils';

					test('something');
				`,
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `describe.skip("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `describe.skip.each([1, 2, 3])("%s", (a, b) => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xdescribe.each([1, 2, 3])("%s", (a, b) => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: "describe[`skip`](\"foo\", function () {})",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `describe["skip"]("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `it.skip("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `it["skip"]("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `test.skip("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: "it.skip.each``(\"foo\", function () {})",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: "test.skip.each``(\"foo\", function () {})",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `it.skip.each([])("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `test.skip.each([])("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `test["skip"]("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xdescribe("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xit("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xtest("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: "xit.each``(\"foo\", function () {})",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: "xtest.each``(\"foo\", function () {})",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xit.each([])("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `xtest.each([])("foo", function () {})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `it("has title but no callback")`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
				},
			},
			{
				Code: `test("has title but no callback")`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
				},
			},
			{
				Code: `it("contains a call to pending", function () { pending() })`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 48},
				},
			},
			{
				Code: `pending();`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `describe("contains a call to pending", function () { pending() })`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 54},
				},
			},
			{
				Code: "import { test } from '@jest/globals';\n\ntest('something');",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 3, Column: 1},
				},
			},
		},
	)
}
